### ISSUE TITLE
wireless doorbell camera feed 

### ISSUE THREAD

**Issue by dpomnean** [fuatakgun/eufy_security#6]
I am trying to get the live feed working for the camera on the wireless 2k doorbell. I was able to add the custom integration and shows all the entities, but when I click on my button card for live feed, the server side shows its not supported. Running Home Assistant Container core-2021.6.2. Installed on Raspberry Pi debian.

`root@worker-2:~/eufy-security-ws# npx ts-node src/bin/server.ts
2021-07-03 04:05:57.705  INFO Eufy Security server listening on port 3000
2021-07-03 04:06:00.213  INFO Connected to station T8010PXXXXXXXX on host 192.168.1.121 and port 21313
2021-07-03 04:06:01.240  INFO Push notification connection successfully established
2021-07-03 04:07:07.631  ERROR Message error
 NotSupportedFeatureError  This functionality is not implemented or supported by T8210XXXXXXXXX
details:
{
  name: 'NotSupportedFeatureError'
}
error stack:
• station.ts:815 setRTSPStream
    node_modules/eufy-security-client/src/http/station.ts:815:19

• message_handler.ts:35 handle
    src/lib/device/message_handler.ts:35:31

• server.ts:46 device
    src/lib/server.ts:46:34

• server.ts:98 receiveMessage
    src/lib/server.ts:98:58

• server.ts:57 <anonymous>
    src/lib/server.ts:57:53

• events.js:375 emit
    events.js:375:28

• domain.js:470 emit
    domain.js:470:12

• websocket.js:834 receiverOnMessage
    node_modules/ws/lib/websocket.js:834:20

• events.js:375 emit
    events.js:375:28

• domain.js:470 emit
    domain.js:470:12`

**Comment by fuatakgun** [comment]
Apparently doorbell doesn't support rtsp, nothing we can do.

**Comment by conorlap** [comment]
Doorbell doesn't support RTSP yeah - although @bropat has gotten video feed from the doorbell via P2P I believe. Any way that can be added to this integration?

**Comment by fuatakgun** [comment]
Hey, this integration is using @bropat's library indirectly, so the error
message actually is coming from there. Can you point me a source about
getting live stream over doorbell, maybe i can implement like that

**Comment by conorlap** [comment]
Hi fuatakgun,
Explanation of setting up live stream from doorbell into HA via IOBroker is here: https://blog.fuzzymistborn.com/controlling-eufy-security-devices-with-iobroker/
It would be great to be able to bypass IOBroker and have the stream directly in the HA integration! 
Thanks for yours and Bropat's work on this so far :)

**Comment by fuatakgun** [comment]
I see now, they have used start live stream rather than start rtsp stream,
let me check with it, will update all of you.

**Comment by fuatakgun** [comment]
I was able to capture stream following that blog post. So, I can get live camera view without needing a RTSP stream, but it required a lot of changes in eufy-security-ws repository.
- Created a simple http file server to serve m3u8 files captured from p2p
- when there is an incoming message over p2p, start recording video locally in a folder
- able to access recorded files over http to serve home asssitant

**Comment by bropat** [comment]
@fuatakgun 

I do not agree with this:

> but it required a lot of changes in eufy-security-ws repository.

eufy-security-ws already has all functionalities for a livestream via P2P:

#### Start live stream

[compatible with schema version: 2+]

```ts
interface {
  messageId: string;
  command: "device.start_livestream";
  serialNumber: string;
}
```

#### Stop live stream

[compatible with schema version: 2+]

```ts
interface {
  messageId: string;
  command: "device.stop_livestream";
  serialNumber: string;
}
```

Events you got:

#### `livestream started`

[compatible with schema version: 2+]

This event is sent whenever the live stream of a device is started.

```ts
interface {
  type: "event";
  event: {
    source: "device";
    event: "livestream started";
    serialNumber: string;
  }
}
```

#### `livestream stopped`

[compatible with schema version: 2+]

This event is sent whenever the live stream of a device is stopped.

```ts
interface {
  type: "event";
  event: {
    source: "device";
    event: "livestream stopped";
    serialNumber: string;
  }
}
```

#### `livestream video data`

[compatible with schema version: 2+]

This event is sent when video data is received during an active live stream of a device. It contains the metadata information and the video chunck.

```ts
interface {
  type: "event";
  event: {
    source: "device";
    event: "livestream video data";
    serialNumber: string;
    buffer: JSONValue,
    metadata: { 
      videoCodec: string;
      videoFPS: number;
      videoHeight: number;
      videoWidth: number;
    }
  }
}
```

#### `livestream audio data`

[compatible with schema version: 2+]

This event is sent when audio data is received during an active live stream of a device. It contains the metadata information and the video chunck.

```ts
interface {
  type: "event";
  event: {
    source: "device";
    event: "livestream audio data";
    serialNumber: string;
    buffer: JSONValue,
    metadata: { 
      audioCodec: string;
    }
  }
}
```

**Comment by fuatakgun** [comment]
Thanks for jumping in but i am not able to convert all these chunks into video stream on python layer and that is because i picked other way. Let me know if you know how to do.

**Comment by bropat** [comment]
Here is an example of how I implemented it in Typescript:

* https://github.com/bropat/ioBroker.eufy-security/blob/master/src/lib/eufy-security/eufy-security.ts#L566-L619
* https://github.com/bropat/ioBroker.eufy-security/blob/0a15e1d125f4fd00144af66d57d8d738140ea619/src/lib/eufy-security/video.ts#L14-L65
* https://github.com/bropat/ioBroker.eufy-security/blob/0a15e1d125f4fd00144af66d57d8d738140ea619/src/lib/eufy-security/video.ts#L99-L168

Basically you have to do the same:

1. pass the video chunks and audio chunks to ffmpeg
2. produce the format you wan't (for example, I have only muxed in the project [ioBroker.eufy-security](https://github.com/bropat/ioBroker.eufy-security)).

I can't tell you how to do it for HomeAssistant in Python, but I think HomeAssistant already has a ffmpeg interface where you can pass the chunks.

**Comment by bropat** [comment]
I've found this example (ha-eufy-security)[https://github.com/nonsleepr/ha-eufy-security].

https://github.com/home-assistant-libs/ha-ffmpeg

**Comment by bropat** [comment]
In fact, this is what @bachya also needs to implement.

**Comment by fuatakgun** [comment]
I had followed your implementation and tried in home assistant but failed.
I believe main reason is converting from chunks to bytes as there is no
known data type as chunks in python. I will try again but if both of hour
clients are going to do this, we can include this functionality onto ws
layer, what do you think?

**Comment by fuatakgun** [comment]
> I've found this example [ha-eufy-security](https://github.com/nonsleepr/ha-eufy-security) using this library [ha-ffmpeg](https://github.com/home-assistant-libs/ha-ffmpeg)

It uses RTSP url to create stream with ffmpeg, not converting buffers back to video

**Comment by bropat** [comment]
Yes I know what I meant to say is that ffmpeg is already available in HomeAssistant.
As I said before the binary data is simply passed to ffmpeg as input via Linux/Unix sockets or Windows pipes.

In the following example `data` is simply a byte array (surely exists in Python too ;)):

```
{
  type: 'event',
  event: {
    source: 'device',
    event: 'livestream video data',
    serialNumber: 'T8410P2021100D6C',
    buffer: {
      type: 'Buffer',
      data: [
        0,
        0,
        0,
        1,
        39,
        100,
        0,
        51,
        173,
        0,
        206,
        128,
        36,
        0,
        163,
        166,
        160,
        32,
        32,
        62,
        0,
        0,
        3,
        0,
        2,
        0,
        0,
        3,
        0,
        60,
        96,
        96,
        0,
        67,
        207,
        0,
        3,
        45,
        187,
        255,
        248,
        20,
        0,
        0,
        0,
        1,
        40,
        238,
        60,
        176,
        0,
        0,
        0,
        1,
        37,
        184,
        32,
        0,
        10,
        91,
        94,
        94,
        26,
        204,
        6,
        214,
        42,
        8,
        109,
        81,
        32,
        212,
        45,
        100,
        219,
        152,
        12,
        70,
        187,
        98,
        124,
        201,
        166,
        41,
        70,
        116,
        22,
        95,
        100,
        114,
        43,
        30,
        106,
        68,
        145,
        73,
        146,
        49,
        48,
        54,
        ... 36837 more items
      ]
    },
    metadata: {
      videoCodec: 'H264',
      videoFPS: 15,
      videoHeight: 1296,
      videoWidth: 2304
    }
  }
}
```

**Comment by alexreddy78** [comment]
> I am trying to get the live feed working for the camera on the wireless 2k doorbell. I was able to add the custom integration and shows all the entities, but when I click on my button card for live feed, the server side shows its not supported. Running Home Assistant Container core-2021.6.2. Installed on Raspberry Pi debian.
> 
> `root@worker-2:~/eufy-security-ws# npx ts-node src/bin/server.ts
> 2021-07-03 04:05:57.705 INFO Eufy Security server listening on port 3000
> 2021-07-03 04:06:00.213 INFO Connected to station T8010PXXXXXXXX on host 192.168.1.121 and port 21313
> 2021-07-03 04:06:01.240 INFO Push notification connection successfully established
> 2021-07-03 04:07:07.631 ERROR Message error
> NotSupportedFeatureError This functionality is not implemented or supported by T8210XXXXXXXXX
> details:
> {
> name: 'NotSupportedFeatureError'
> }
> error stack:
> • station.ts:815 setRTSPStream
> node_modules/eufy-security-client/src/http/station.ts:815:19
> 
> • message_handler.ts:35 handle
> src/lib/device/message_handler.ts:35:31
> 
> • server.ts:46 device
> src/lib/server.ts:46:34
> 
> • server.ts:98 receiveMessage
> src/lib/server.ts:98:58
> 
> • server.ts:57 
> src/lib/server.ts:57:53
> 
> • events.js:375 emit
> events.js:375:28
> 
> • domain.js:470 emit
> domain.js:470:12
> 
> • websocket.js:834 receiverOnMessage
> node_modules/ws/lib/websocket.js:834:20
> 
> • events.js:375 emit
> events.js:375:28
> 
> • domain.js:470 emit
> domain.js:470:12`

Dear Dpomnean,

Can you pls advise if possible to take snapshot in HA, from doorbell upon move detection? Is it react immediately or need some time?

BRGDS

**Comment by fuatakgun** [comment]
I was able to generate video stream locally, thanks for guidance, on the other hand, there us at least 10 seconds difference between stream and current time. How is the performance on iobroker, @bropat ?

**Comment by fuatakgun** [comment]
> > I am trying to get the live feed working for the camera on the wireless 2k doorbell. I was able to add the custom integration and shows all the entities, but when I click on my button card for live feed, the server side shows its not supported. Running Home Assistant Container core-2021.6.2. Installed on Raspberry Pi debian.
> > 
> > `root@worker-2:~/eufy-security-ws# npx ts-node src/bin/server.ts
> > 2021-07-03 04:05:57.705 INFO Eufy Security server listening on port 3000
> > 2021-07-03 04:06:00.213 INFO Connected to station T8010PXXXXXXXX on host 192.168.1.121 and port 21313
> > 2021-07-03 04:06:01.240 INFO Push notification connection successfully established
> > 2021-07-03 04:07:07.631 ERROR Message error
> > NotSupportedFeatureError This functionality is not implemented or supported by T8210XXXXXXXXX
> > details:
> > {
> > name: 'NotSupportedFeatureError'
> > }
> > error stack:
> > • station.ts:815 setRTSPStream
> > node_modules/eufy-security-client/src/http/station.ts:815:19
> > 
> > • message_handler.ts:35 handle
> > src/lib/device/message_handler.ts:35:31
> > 
> > • server.ts:46 device
> > src/lib/server.ts:46:34
> > 
> > • server.ts:98 receiveMessage
> > src/lib/server.ts:98:58
> > 
> > • server.ts:57 
> > src/lib/server.ts:57:53
> > 
> > • events.js:375 emit
> > events.js:375:28
> > 
> > • domain.js:470 emit
> > domain.js:470:12
> > 
> > • websocket.js:834 receiverOnMessage
> > node_modules/ws/lib/websocket.js:834:20
> > 
> > • events.js:375 emit
> > events.js:375:28
> > 
> > • domain.js:470 emit
> > domain.js:470:12`
> 
> Dear Dpomnean,
> 
> Can you pls advise if possible to take snapshot in HA, from doorbell upon move detection? Is it react immediately or need some time?
> 
> BRGDS
> 

Event based snapshot capturing should be fast BUT current eufy-client-ws is not publishing latest picture URL with motion event. I have created a pull request to fix it. Until then, you can modify it locally to get latest picture url quickly. 

Related pr: https://github.com/bropat/eufy-security-client/pull/20

**Comment by dpomnean** [comment]
I also contacted Eufy if they support RTSP feed for the doorbell camera and they said no, and asked if it will in the future. I have not gotten a reply back from them so I feel that support may not happen.

**Comment by conorlap** [comment]
> I also contacted Eufy if they support RTSP feed for the doorbell camera and they said no, and asked if it will in the future. I have not gotten a reply back from them so I feel that support may not happen.

Yeah there are a lot of users begging for RTSP on the doorbell, namely on the [Eufy Forums. ](https://communitysecurity.eufylife.com/t/rtsp-for-eufy-doorbell-2k/339139) 
I've also been encouraging people to keep emailing Eufy to ask when this feature is coming as they did indicate over a year back that it **would** be coming "soon". Not holding my breath though, but we can try!

**Comment by fuatakgun** [comment]
So, as a fact, we know that it doesn't support RTSP endpoint and we are on our own. Thanks to super work by @bropat we can get the video but my experience with eufy cam 2c is super laggy.

It can be because of slow processing in home assistant or a lot of hop between source and destination.

Currently, in my local setup, i am able to live stream ( NOT RTSP) but with 10 seconds delay.

I will publish my changes and you can share your experience.

**Comment by dpomnean** [comment]
@fuatakgun did you merge your changes?

**Comment by fuatakgun** [comment]
I will merge today, but it is still flaky, has at least 10 seconds delay.

**Comment by fuatakgun** [comment]
> I was able to generate video stream locally, thanks for guidance, on the other hand, there us at least 10 seconds difference between stream and current time. How is the performance on iobroker, @bropat ?

@bropat any feedback is welcome here :-)

**Comment by dpomnean** [comment]
@fuatakgun I will give it a try and see how it performs. Could be dependent on a few factors

**Comment by fuatakgun** [comment]
> @fuatakgun I will give it a try and see how it performs. Could be dependent on a few factors

Might be, i am using an old android phone as home assistant server, maybe it is the bottleneck when converting from data chunks to video stream

**Comment by dpomnean** [comment]
If you don't want it in the master, you could create a second branch for testing and I can clone it and let you know how it goes

**Comment by mollux** [comment]
I have also a setup with home assistant with an Eufy wired doorbell, so I can test it out too.

**Comment by fuatakgun** [comment]
please test on this branch: https://github.com/fuatakgun/eufy_security/tree/livestream
update your HACS entry with url above and you will receive latest changes.
**WARNING AHEAD: camera.turn_on service will check RTSP support and fallback to livestream if needed. for live streaming, it will create a test.m3u8 stream file and many segments file in main folder, so you can have many temp files. I am still working on a fix to remove these or find a better place to store.**
While streaming is enabled, camera picture will be received from stream.

**Comment by dpomnean** [comment]
@fuatakgun So I cloned the branch, confirmed it had the changes in some of your python files, restarted home assistant and removed, and added the custom integration for the livestream. Doesn't seem to be falling back to the livestream and using RTSP, unless I'm doing something wrong?

my card:

```type: grid
cards:
  - type: button
    tap_action:
      action: call-service
      service: camera.turn_on
      service_data: {}
      target:
        entity_id: camera.doorbell
    entity: camera.doorbell
    show_state: true
    show_icon: false
    name: toogle doorbell
    hold_action:
      action: call-service
      service: camera.turn_off
      service_data: {}
      target:
        entity_id: camera.doorbell
```


Getting this error from eufy-security-ws

```
2021-07-18 21:31:23.123  ERROR Message error
 NotSupportedFeatureError  This functionality is not implemented or supported by T8210P00204107B0
details:
{
  name: 'NotSupportedFeatureError'
}
error stack:
• station.ts:815 setRTSPStream
    node_modules/eufy-security-client/src/http/station.ts:815:19

• message_handler.ts:35 handle
    src/lib/device/message_handler.ts:35:31

• server.ts:46 device
    src/lib/server.ts:46:34

• server.ts:98 receiveMessage
    src/lib/server.ts:98:58

• server.ts:57 <anonymous>
    src/lib/server.ts:57:53

• events.js:375 emit
    events.js:375:28

• domain.js:470 emit
    domain.js:470:12

• websocket.js:834 receiverOnMessage
    node_modules/ws/lib/websocket.js:834:20

• events.js:375 emit
    events.js:375:28

• domain.js:470 emit
    domain.js:470:12
```

**Comment by fuatakgun** [comment]
Can you call start live stream service?

**Comment by dpomnean** [comment]
Ah, tried calling the service and got this error from the HA logs

```
Logger: homeassistant.components.websocket_api.http.connection
Source: custom_components/eufy_security/__init__.py:33
Integration: Home Assistant WebSocket API (documentation, issues)
First occurred: 6:03:34 PM (8 occurrences)
Last logged: 6:05:51 PM

[2809359304] 'serialNumber'
[2809359304] Error handling message: Unknown error
[2836203864] 'serialNumber'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 185, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1491, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1526, in _execute_service
    await handler.job.target(service_call)
  File "/config/custom_components/eufy_security/__init__.py", line 33, in async_handle_start_livestream
    serial_number = call.data["serialNumber"]
KeyError: 'serialNumber'
```

**Comment by fuatakgun** [comment]
of course, my bad, you need to send serial number of target camera
```
service: eufy_security.stop_livestream
data:
  serialNumber: "your_serial_number_goes_here"
```

**Comment by dpomnean** [comment]
Added that in for the service and got a warning this time, so it does seem like its trying not to use RTSP

```
Logger: homeassistant.helpers.service
Source: helpers/service.py:414
First occurred: 8:20:32 PM (2 occurrences)
Last logged: 8:20:32 PM

Unable to find services.yaml for the eufy_security integration
```


And then I get this error from home assistant container

```
2021-07-18 20:11:29 ERROR (MainThread) [homeassistant.components.camera] Error requesting stream: camera.doorbell does not support play stream service
```

**Comment by fuatakgun** [comment]
Please keep us updated @dpomnean , i have received an email notification but comment is not here anymore.

**Comment by dpomnean** [comment]
@fuatakgun livestream works fine from the eufy app. I was confirming the serial number was the same as what HA was reporting. But streaming from HA still wasn't working for me. I have the battery doorbell, not sure if that plays a factor.

**Comment by fuatakgun** [comment]
could you try to call service over developer page (example below)? moreover, are you using latest version of `eufy-security-ws`? I recently upgraded it and had issues with live video stream so I had to roll back. I am using this forked version which does not include last 2 commits.  https://github.com/fuatakgun/eufy-security-ws . I am tracking that issue over here: https://github.com/bropat/eufy-security-client/issues/23

```
service: eufy_security.start_livestream
data:
  serialNumber: "your_serial_number_goes_here"
```

**Comment by dpomnean** [comment]
Ah, I am using the latest that was released. I will test with your fork. And yes I also tried with the service page, that ended up crashing my HA instance and gave me this error. But I will try your fork and see

```
2021-07-18 20:11:29 ERROR (MainThread) [homeassistant.components.camera] Error requesting stream: camera.doorbell does not support play stream service
```

**Comment by fuatakgun** [comment]
That is just a warning nothing important, because of missing services.yaml, it is not able to render it on ui in development page. On the other hand, i am worried about crashing, can you save your home assistant log before restarting it so we can take a look at it.

**Comment by bropat** [comment]
> I was able to generate video stream locally, thanks for guidance, on the other hand, there us at least 10 seconds difference between stream and current time. How is the performance on iobroker, @bropat ?

The best you can do, to my knowledge, is about 6 seconds with normal HLS. If you want to achieve more you have to resort to low latency streaming protocols.

**Comment by dpomnean** [comment]
@fuatakgun Ok, so I don't think it was crashing before, I think it was just hanging so after restarting the docker container. I went into service and start the livestream and nothing happens or shows up, nothing in logs. I go to click on it again and tells me this. Was I supposed to have a popup window for the stream?

```
code frame:
  069 |                         event: DeviceEvent.livestreamStarted,
  070 |                         serialNumber: serialNumber,
  071 |                     })
  072 |                     client.receiveLivestream[device.getSerial()] = true;
  073 |                 } else {
> 074 |                     throw new LivestreamAlreadyRunningError(`Livestream for device ${device.getSerial()} is already running`);
                                  ^
  075 |                 }
  076 |                 return { };
  077 |             case DeviceCommand.stopLivestream:
  078 |                 if (client.receiveLivestream[device.getSerial()] !== true) {
  079 |                     throw new LivestreamNotRunningError(`Start of livestream for device ${device.getSerial()} was not also requested by this client`);
```

**Comment by fuatakgun** [comment]
After starting live stream, click on camera image on frontend, a popup will open and you can see livestream with some latency

**Comment by dpomnean** [comment]
I think were making progress. Started the service, made sure it was running, clicked on the picture entity image for the stream, it did open a popup window, but HA threw this error. Nothing in the eufy-security-ws

```
2021-07-19 21:05:59 ERROR (stream_worker) [homeassistant.components.stream.worker] Error opening stream test.m3u8
```

**Comment by conorlap** [comment]
I can't see the eufy_security.start_livestream service, all I have is 'eufy_security.send_message' under services. 
I've installed this branch: https://github.com/fuatakgun/eufy_security/tree/livestream via HACS and I'm running the very latest bropat/eufy-security-ws

**Comment by fuatakgun** [comment]
Please update eufy_security integration, it is updated 4-5 days ago.

**Comment by fuatakgun** [comment]
Merged into mainline, you can try to update it now.

**Comment by dpomnean** [comment]
I had to do this and replace it with the custom integrations folder for eufy_security

```git clone -b livestream https://github.com/fuatakgun/eufy_security```

**Comment by dpomnean** [comment]
@fuatakgun did you see my comment above about 'Error opening stream test.m3u8'

**Comment by fuatakgun** [comment]
> @fuatakgun did you see my comment above about 'Error opening stream test.m3u8'

When you call start livestream service, camera should write 'streaming' on ui. After that if you click and wait around 10 seconds (that is the problematic part) you should get a file playing, can you retest?

One last thing, when you start streaming, check home assistant main folder, you should see test.m3u8 and ffmpeg-{some date}.log file. Can you share log file content if you don't get video?

**Comment by dpomnean** [comment]
Yeah I did that, the card did say it was 'streaming'. I clicked clicked on the button and popped up a window of a video screen but kept spinning waiting for a good amount of time. I did see it write some ffmpeg files so when I get home I will see if it creates that test file and a new log file.

**Comment by fuatakgun** [comment]
Either it had captured some error in ffmpeg logs about incoming data or it should have created test.m3u8 file, keep us updated.

It would be nice to have more testers on this, all is working for me with 2c camera but we are missing different data points. I feel like i have to buy a battery doorbell eventually :-)

**Comment by dpomnean** [comment]
Here are my steps. Stop eufy-security-ws. Start eufy-security-ws. Restart HA container. Start the livestream service and go back to lovelace. (button now shows 'streaming'). clicked on the picture and video pop up started, didnt play anything and produced this log in the browser, then froze up my entire HA instance, and said
```Error opening stream test.m3u8```

No test.m3u8 file was created (I even tried manually creating that file), but the ffmpeg log file was created.
[ffmpeg-20210720-190600.log](https://github.com/fuatakgun/eufy_security/files/6851699/ffmpeg-20210720-190600.log)


![image](https://user-images.githubusercontent.com/17032114/126406619-e27f5e6e-7380-4dae-84fc-aaf7d4b00fcc.png)

**Comment by bropat** [comment]
@fuatakgun The video stream can come to way with H264 or H265. Depending on the device, both can be supported, such as with the Battery Doorbell (adjustable in the settings!).
Here you have to query the video metadata, property `videoCodec`:

https://github.com/bropat/eufy-security-client/blob/e398105a0cbffe23a373eda6779c1338d9535455/src/p2p/types.ts#L594-L598

Here you have to check the value and pass the correct codec to ffmpeg depending on it.

**Comment by fuatakgun** [comment]
You can see in your logs that ffmpeg is throwing errors related to old
version.

Update your FFmpeg version to the newest one from Git. If the problem
still occurs, it means that your file has a feature which has not been
implemented.

**Comment by fuatakgun** [comment]
This is a valid call-out, i have missed this, let me fix it.

**Comment by fuatakgun** [comment]
@bropat , i have one question regarding to finding out correct streaming option between `rtsp` or `livestream`, is there a way to find out what are supported for the camera?

for my 2c camera, both options are supported but battery doorbell supports only livestream, how can i find it out?

**Comment by bropat** [comment]
@fuatakgun I would suggest using the `hasProperty` function of the `Device` class or subclasses to check the `PropertyName.DeviceRTSPStream` property for existence.
I have tried to map the properties to the respective devices here:

https://github.com/bropat/eufy-security-client/blob/e398105a0cbffe23a373eda6779c1338d9535455/src/http/types.ts#L1395-L2179

Not everything is 100% mapped yet. If you find something missing or incorrect just report it.

However, to find out if a device supports the livestream there is also `hasCommand` with the following device mappings:

https://github.com/bropat/eufy-security-client/blob/e398105a0cbffe23a373eda6779c1338d9535455/src/http/types.ts#L2536-L2654

Specifically, you can check for `CommandName.DeviceStartLivestream`.

**Comment by fuatakgun** [comment]
@dpomnean please reach out to me over email fuatakgun@gmail.com and we can work closely today (tonight in my timezone :) )

**Comment by fuatakgun** [comment]
> CommandName.DeviceStartLivestr

if this is not exposed through web socket, I will not be able to do it, right?

**Comment by dpomnean** [comment]
I will be able to help around 10pm EST tonight, I will email you

**Comment by bropat** [comment]
> > CommandName.DeviceStartLivestream
> 
> if this is not exposed through web socket, I will not be able to do it, right?

You are right, this part has not been provided in `eufy-security-ws` yet :P

**Comment by fuatakgun** [comment]
> > > CommandName.DeviceStartLivestream
> > 
> > 
> > if this is not exposed through web socket, I will not be able to do it, right?
> 
> You are right, this part has not been provided in `eufy-security-ws` yet :P

so, my intention was to enable rtsp (if supported), otherwise to enable live stream. 
do you think you can add this functionality to WS anytime? until then, I can ask users to go with explicit service call approach

**Comment by fuatakgun** [comment]
Please take a look at README file, I have updated the repository and created the add-on for docker usage. no more manual setup is required.

**Comment by dpomnean** [comment]
Hey there. I downloaded your latest changes. I have to clone it into the custom_components since I have the docker install. I can write up how I set it up later. But this is the latest.

I went through the normal steps, starting the service, clicking on the stream picture and got this error message from HA.

```
2021-07-21 20:56:00 ERROR (MainThread)
[custom_components.eufy_security] eufy_security Exception in video :
I/O operation on closed file - traceback: Traceback (most recent call
last):
  File "/config/custom_components/eufy_security/camera.py", line 227,
in process_frames
    self.ffmpeg.process.stdin.write(frame_bytes)
ValueError: I/O operation on closed file
```

and then I got this error from the browser console but only twice, not
repeating like it was before.

```
GET http://192.168.1.65:8123/api/hls/275f2fd06089100d440f93654d99d42d455c48eae078029208851877eebf344d/master_playlist.m3u8
 404 (Not Found)
```

I also had this in the ffmpeg log file at the bottom

```
Unknown input format: 'h265'
```


If I set the 'Video Encoding Format' on the eufy doorbell to 'low compression quality video coding format' the livestream will work! But got the errors mentioned above when using the 'High' compression. I noticed it still freezes up the service after a certain amount of time. Even stopping and starting the service doesn't fix it. Does the live feed have a limit on how long it can run?

**Comment by dpomnean** [comment]
Updated my card to reflect the new service. Seems like it working fine on the low quality video coding format. The stream stopped working around 3 and a half minutes, but I closed out the stream and reopened and started working fine again.

```
type: grid
cards:
  - type: button
    tap_action:
      action: call-service
      service: eufy_security.start_livestream
      target:
        device_id: insert_id_here
    show_state: true
    show_icon: false
    name: toggle doorbell
    hold_action:
      action: call-service
      service: eufy_security.stop_livestream
      target:
        device_id: insert_id_here

```

**Comment by fuatakgun** [comment]
I would like to resolve this issue as it became gigantic. Please reopen if your believe you are having issues with live stream.

**Comment by bropat** [comment]
> do you think you can add this functionality to WS anytime?

@fuatakgun Added in version 0.3.3 released a few seconds ago
